mitiq/folding_cirq.py

```diff
@@ -1,76 +1,197 @@
+    """Returns a new circuit with moments folded by mapping
+
+    M_i -> M_i M_i^dag M_i
+
+    where M_i is a moment specified by an integer in moment_indices.
+
+    Args:
+        circuit: Circuit to apply folding operation to.
+        moment_indices: List of integers that specify moments to fold.
+    """
+    folded = deepcopy(circuit)
+    shift = 0
+    for i in moment_indices:
+        folded.insert(i + shift, [inverse(circuit[i]), circuit[i]])
+        shift += 2
+    return folded
 
 
-def sampling_stretcher(circuit: Circuit, stretch: float) -> Circuit:
-    """Applies the map G -> G G^dag G to a random subset of gates 
-    of the input circuit.
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number within 1 and 3."""
+def fold_gates_from_left(circuit: Circuit, stretch: float) -> Circuit:
+    """Returns a new folded circuit by applying the map G -> G G^dag G to a subset of gates  of the input circuit.
 
-    if not ((stretch >= 1) and (stretch <= 3)):
-        raise ValueError("The stretch factor must be a real number within 1 and 3.")
+    The folded circuit has a number of gates approximately equal to stretch * n where n is the number of gates in
+    the input circuit.
+
+    Args:
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by. Any real number in the interval [1, 3].
 
-    depth = len(circuit)
-    reduced_depth = int(depth * (stretch - 1) / 2)
-    sub_indices = random.sample(range(depth), reduced_depth)
-    return fold_gates(circuit, sub_indices)
+    Note:
+        Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
+    """
+    folded = deepcopy(circuit)
 
+    if np.isclose(stretch, 1.0, atol=1e-2):
+        return folded
 
-def left_stretcher(circuit: Circuit, stretch: float) -> Circuit:
-    """Applies the map G -> G G^dag G to a subset of gates of the input circuit (first part of it).
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number within 1 and 3."""
+    if not 1 < stretch <= 3:
+        raise ValueError("The stretch factor must be a real number between 1 and 3.")
 
-    if not ((stretch >= 1) and (stretch <= 3)):
-        raise ValueError("The stretch factor must be a real number within 1 and 3.")
+    ngates = len(list(folded.all_operations()))
+    num_to_fold = int(ngates * (stretch - 1.0) / 2.0)
+    num_folded = 0
+    moment_shift = 0
 
-    depth = len(circuit)
-    reduced_depth = int(depth * (stretch - 1) / 2)
-    sub_indices = list(range(reduced_depth))
-    return fold_gates(circuit, sub_indices)
+    for (moment_index, moment) in enumerate(circuit):
+        for gate_index in range(len(moment)):
+            # TODO: It could be expensive to call fold_gate...(...) which makes a deepcopy of the circuit each call.
```

Comment:
I think this will work. I wonder if we would ever require `mutate=False`, certainly I think users wouldn't need this. It might make sense to define this as a private function

```python
def _fold_gate_at_index_in_moment(circuit: Circuit, moment_index: int, gate_index: int) -> Circuit:
    op = circuit[moment_index].operations[gate_index]
    circuit.insert(moment_index, [inverse(op), op], strategy=InsertStrategy.NEW)
```

Reply:
+1 to your suggestion above